torch/csrc/autograd/python_variable_numpy.cpp

```diff
@@ -0,0 +1,107 @@
+#include "python_variable_numpy.h"
+
+#ifndef WITH_NUMPY
+namespace torch { namespace autograd {
+PyObject * THPVariable_numpy(PyObject* pyself, PyObject* arg) {
+  return PyErr_Format(PyExc_RuntimeError, "PyTorch was compiled without NumPy support");
+}
+}}
+#else
+
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/autograd/python_variable.h"
+
+#include <ATen/ATen.h>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+
+#define NO_IMPORT_ARRAY
+#define PY_ARRAY_UNIQUE_SYMBOL __numpy_array_api
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include <numpy/arrayobject.h>
+
+using namespace at;
+
+namespace torch { namespace autograd {
+
+static std::vector<npy_intp> cast_numpy(IntList x) {
+  auto nelem = x.size();
+  auto result = std::vector<npy_intp>(nelem);
+  for (size_t i = 0; i < nelem; i++) {
+    result[i] = static_cast<npy_intp>(x[i]);
+  }
+  return result;
+}
+
+static int numpy_dtype(const at::Type& type);
+
+PyObject * THPVariable_numpy(PyObject* pyself, PyObject* arg) {
```

Comment:
I'm wondering if we should allow converting Variables that require grad to numpy. It might make it easy for people to shoot themselves in the foot because they will think that we can also differentiate numpy code, which isn't true.

Reply:
Eh... that path starts getting pretty messy. Do we disallow `storage()`, `tolist()`, `data_ptr()`, pickling, and copying via `copy` module on variables that require grad?

Reply:
I think `numpy()` is slightly different. There are not tools that differentiate data pointers or list operation, but there are some for numpy (HIPS autograd). This is the place where both supporting and not supporting differentiation makes sense, and you have no idea what happens unless you check.

Reply:
I'll make it an error. We can always allow it later if it turns out to be too annoying.